Modules: A/B testing support

postindustria-code opened the pull request:
Addresses Issue #4527 
                                                                                                                                                                                                                                      
**A/B Testing Support for Modules**
                                                                                                                                                                                                                                                    
Adds A/B testing support to the hook execution system, allowing publishers to configure traffic splits across module execution plans.

postindustria-code committed:
A/B testing support for modules

postindustria-code committed:
Merge branch 'master' into ab-testing-support-for-modules

# Conflicts:
#	hooks/hookexecution/executor_test.go
#	hooks/hookexecution/mocks_test.go

postindustria-code committed:
[Issue 4527] - Add A/B test targeting keyword support to auction responses

postindustria-code committed:
[Issue 3681] - Add A/B test targeting keyword support to auction responses

postindustria-code committed:
Merge remote-tracking branch 'origin/ab-testing-support-for-modules' into ab-testing-support-for-modules

postindustria-code committed:
Merge branch 'master' into ab-testing-support-for-modules

postindustria-code committed:
Merge branch 'master' into ab-testing-support-for-modules

justadreamer commented:
@postindustria-code please see a slight change in the spec for the GAM keyword logic (the value becomes a string rather than 0/1): https://github.com/prebid/prebid-server/issues/4527#issuecomment-3426933880

antoxas1986 reviewed (commented):
## Review: Modules: A/B testing support

### Critical

**1. Data race on `ABTests` state (`hooks/hookexecution/abtests.go`)**

`planLoaded`, `runMap`, `logMap`, and `targetingMap` are read and written without synchronization. The existing `hookExecutor` runs `ExecuteBidderRequestStage` and `ExecuteRawBidderResponseStage` concurrently across bidders — both call `applyABTestPlan` → `ab.Run()` — which reads and writes these fields with no lock held. The mutex only protects `loggedMap`. A `go test -race` run with multiple bidders will surface this.

**Suggested fix:** Use `sync.Once` wrapping `planHost()` + `planAccount()`, removing the `planLoaded` flag entirely:

```go
type ABTests struct {
    ...
    initOnce sync.Once
    // remove planLoaded field
}

func (t *ABTests) init() {
    t.initOnce.Do(func() {
        t.planHost()
        t.planAccount()
    })
}

func (t *ABTests) Run(module string) bool {
    t.init()
    t.mu.RLock()
    defer t.mu.RUnlock()
    val, ok := t.runMap[module]
    if !ok {
        return true
    }
    return val
}
```

All map writes in `planHost`/`planAccount` happen inside `initOnce.Do` (single-threaded), and all subsequent reads use `sync.RWMutex`.

---

**2. Five wrong `Stage` (and `Entity`) constants in empty-plan branches (`hooks/hookexecution/executor.go`)**

When `applyABTestPlan` reduces the plan to empty, five methods emit a `StageOutcome` with a hardcoded `Stage: hooks.StageEntrypoint.String()`. This corrupts analytics — downstream consumers see phantom entrypoint records for every skipped stage.

| Method | Stage emitted (wrong) | Should be | Entity emitted (wrong) | Should be |
|---|---|---|---|---|
| `ExecuteProcessedAuctionStage` | `StageEntrypoint` | `StageProcessedAuctionRequest` | `entityHttpRequest` | `entityAuctionRequest` |
| `ExecuteBidderRequestStage` | `StageEntrypoint` | `StageBidderRequest` | `entityHttpRequest` | bidder-scoped |
| `ExecuteRawBidderResponseStage` | `StageEntrypoint` | `StageRawBidderResponse` | `entityHttpRequest` | bidder-scoped |
| `ExecuteAllProcessedBidResponsesStage` | `StageEntrypoint` | `StageAllProcessedBidResponses` | `entityHttpRequest` | correct entity |
| `ExecuteAuctionResponseStage` | `StageEntrypoint` | `StageAuctionResponse` | `entityHttpRequest` | correct entity |

**Suggested fix:** Extract a constant (or local `stageName` variable) at the top of each method, matching the pattern used in the non-empty path. For example in `ExecuteProcessedAuctionStage`:

```go
const stageName = hooks.StageProcessedAuctionRequest.String()

if len(plan) == 0 {
    outcome := StageOutcome{
        Entity: entityAuctionRequest,
        Stage:  stageName,
    }
    ...
}
```

Also add test assertions on `Stage` and `Entity` in the empty-plan outcome to prevent regression.

---

### High

**3. Spec compliance — targeting value must be a string, not `"0"`/`"1"` (existing PR comment)**

Per @justadreamer's comment referencing the spec update in #4527, `GetTargetingKeywords()` must not emit `"1"`/`"0"`. The constants `ResultStatusRun = "run"` and `ResultStatusSkip = "skip"` already exist in `hooks/hookanalytics/analytics.go` — added in this same PR. Using them makes analytics and GAM targeting representations consistent.

**Suggested fix:**

```go
// hooks/hookexecution/abtests.go
import "github.com/prebid/prebid-server/v3/hooks/hookanalytics"

func (t *ABTests) GetTargetingKeywords() map[string]string {
    t.init()
    result := make(map[string]string)
    for module, keyword := range t.targetingMap {
        if keyword != "" {
            value := string(hookanalytics.ResultStatusSkip)
            if t.runMap[module] {
                value = string(hookanalytics.ResultStatusRun)
            }
            result[keyword] = value
        }
    }
    return result
}
```

Also update the `config/hooks.go` struct comment on `AdServerTargeting` to remove the `"1"`/`"0"` reference.

---

**4. `SetAccountID` ignores app traffic (`hooks/hookexecution/abtests.go`)**

`SetAccountID` only reads `site.publisher.id`. App/CTV/DOOH requests use `app.publisher.id`, so account-scoped A/B tests silently skip all app traffic.

**Suggested fix:**

```go
func (t *ABTests) SetAccountID(body []byte) {
    t.accountID = gjson.GetBytes(body, "site.publisher.id").String()
    if t.accountID == "" {
        t.accountID = gjson.GetBytes(body, "app.publisher.id").String()
    }
}
```

---

**5. `ExecuteExitpointStage` not wired for A/B tests (`hooks/hookexecution/executor.go`)**

Every stage method was updated to call `applyABTestPlan` and `WriteOutcome` — except `ExecuteExitpointStage`. A module participating in both A/B testing and the exitpoint stage will run unconditionally there regardless of the A/B decision, and no analytics tag will be written.

**Suggested fix:** Apply the same pattern used in all other stage methods to `ExecuteExitpointStage`.

---

### Medium

**6. `logMap` not cleaned on account-level `Enabled: false` (`hooks/hookexecution/abtests.go`)**

When `planAccount()` disables a module, it deletes from `runMap` and `targetingMap` but leaves `logMap` intact. If the host plan had `LogAnalyticsTag: true`, `WriteOutcome` will still emit analytics tags for the disabled module.

**Suggested fix:** Add `delete(t.logMap, abtest.ModuleCode)` alongside the other two deletes:

```go
if abtest.Enabled == nil || !*abtest.Enabled {
    delete(t.runMap, abtest.ModuleCode)
    delete(t.targetingMap, abtest.ModuleCode)
    delete(t.logMap, abtest.ModuleCode)
    continue
}
```

---

**7. `givenAccountConfiguration` in tests has no effect (`hooks/hookexecution/executor_test.go`)**

The test loop hardcodes `exec.SetAccount(&config.Account{ID: "account-id"})` instead of reading `test.givenAccountConfiguration`. Test cases that set this field to verify account-level overrides are not actually exercising that path — they give false confidence.

**Suggested fix:**

```go
account := test.givenAccountConfiguration
if account == nil {
    account = &config.Account{ID: "account-id"}
}
exec.SetAccount(account)
```

---

**8. Test regression — existing `TestExecuteRawAuctionStage` cases removed**

The PR restructured `TestExecuteRawAuctionStage` and removed the "Request changed if hooks return mutations" and "Stage execution can be rejected" test cases. The raw auction stage operates on `[]byte` bodies (unlike `ProcessedAuctionStage` which uses `RequestWrapper`) and the removed cases covered meaningfully different behavior. Please restore them.

---

**9. `PercentActive` values > 100 not validated**

`PercentActive` is `*uint16`, allowing values up to 65535. The guard `*abtest.PercentActive < uint16(100)` silently treats any value ≥ 100 as 100% active. 

**Suggested fix:** Add config validation (e.g., in a `Validate()` method or in the existing config validation pass) to enforce the `[0, 100]` range and return an error for out-of-range values.

---

### Design

**10. `ABTests` should be an interface**

`hookExecutor.abTests` is typed as `*ABTests` (concrete). This prevents mocking in tests, is inconsistent with the `HookStageExecutor` interface pattern, and would panic if `nil` is passed to `NewHookExecutor`.

**Suggested fix:** Define an `ABTester` interface:

```go
type ABTester interface {
    SetAccount(account *config.Account)
    SetAccountID(body []byte)
    Run(module string) bool
    WriteOutcome(outcome *StageOutcome)
    GetTargetingKeywords() map[string]string
}
```

`EmptyHookExecutor` can hold a no-op implementation. `NewHookExecutor` accepts `ABTester`.

---

**11. `applyABTestPlan` should move to `abtests.go`**

The function only operates on `*ABTests` but lives in `executor.go`. Moving it as a method `func (t *ABTests) FilterPlan[T any](plan hooks.Plan[T]) hooks.Plan[T]` consolidates all A/B test logic.

---

**12. Duplicated logic between `planHost` and `planAccount`**

The two functions share nearly identical loop bodies (resolving `logMap`, `targetingMap`, `runMap`). The only differences are account filtering and delete-vs-skip on disable.

**Suggested fix:** Extract a private helper:

```go
func (t *ABTests) applyPlan(tests []config.ABTest, deleteOnDisable bool) {
    for _, abtest := range tests {
        ...
    }
}
```

---

**13. `DefaultAccountExecutionPlan` fully bypassed when account is set**

`planAccount()` ignores `DefaultAccountExecutionPlan` entirely when `t.account != nil`. This differs from how the regular execution plan works (where the default applies when no account-specific plan exists). If intentional, please document it explicitly in a comment — otherwise this is a configuration layering bug.

---

### Minor

**14. Empty `Action` in synthesized skipped-module `HookOutcome`**

In `WriteOutcome`, the synthetic `HookOutcome` for a skipped module uses `Action: ""`. Real invocations use `ActionNone = "no_action"`. Use `ActionNone` for consistency in analytics output.

**15. Per-request `glog.Warning` for missing `module_code` should be startup validation**

The warning for a missing `ModuleCode` fires on every request for misconfigured entries. Move this check to the config validation pass so it surfaces once at startup.

postindustria-code committed:
Merge branch 'master' into ab-testing-support-for-modules

postindustria-code committed:
fix versioning

postindustria-code committed:
fix log import

scr-oath commented on endpoints/openrtb2/auction.go:
```diff
@@ -399,6 +405,38 @@ func sendAuctionResponse(
+	}
+
+	var respExt openrtb_ext.ExtBidResponse
+	if err := jsonutil.Unmarshal(response.Ext, &respExt); err != nil {
```
question: AMP path (amp_auction.go line 401) has a collision guard (`if _, exists := targets[keyword]; !exists`), but this auction path writes unconditionally. Is the asymmetry intentional? A config-supplied keyword could overwrite standard prebid targeting keys here but not in AMP.

scr-oath reviewed (changes requested):
## Review: Modules A/B Testing Support

**The bottom line:** This feature is valuable but has several critical bugs that need fixing before merge.

### Blockers

| # | Severity | File | Issue |
|---|----------|------|-------|
| 1 | Critical bug | executor.go:184,233,272,313,349 | Wrong stage name in 5 of 7 empty-plan branches (copy-paste of `StageEntrypoint`) |
| 2 | Critical race | abtests.go:48-60 | `Run()` has unprotected lazy init + concurrent map reads; will panic under load |
| 3 | Critical logic | abtests.go:150-188 | `planAccount()` ignores `Accounts` filter (unlike `planHost()`) |
| 4 | Correctness | abtests.go:218-223 | `GetTargetingKeywords()` duplicates lazy-load — can diverge from `Run()` decisions |
| 5 | Race | abtests.go:62-107 | TOCTOU on `loggedMap` in `WriteOutcome` — can produce duplicate analytics |
| 6 | Bug | abtests.go:44-46 | `SetAccountID` only checks `site.publisher.id`, misses `app` and `dooh` traffic |

### Non-blocking

| # | Severity | File | Issue |
|---|----------|------|-------|
| 7 | Medium | abtests.go | 8 exported identifiers missing godoc; non-idiomatic `type()` grouping |
| 8 | Medium | auction.go vs amp_auction.go | Targeting keyword collision guard inconsistent between endpoints |
| 9 | Low | abtests.go:115 | Uses `glog.Warning` instead of project's `logger` package |

### golangci-lint findings

Installed `golangci-lint` v2.11.3 via `brew tap golangci/tap && brew install golangci/tap/golangci-lint`. Default linters found 1 issue in changed code:
- `hooks/hookexecution/executor.go:296` — `ineffassign`: assignment to `response` is never used

With `revive` enabled, 8 missing-godoc violations in `abtests.go` and 12+ in `executor.go` (most pre-existing).

### Process feedback

**suggestion:** The PR title is interesting but the body is empty. Consider filing a GitHub Issue to document the problem statement, goals, and expected impact of A/B testing for modules, then reference it in the PR title (e.g., `Modules: A/B testing support (fixes #NNNN)`). This helps reviewers understand the "why" and gives future contributors a place to find design decisions.

postindustria-code committed:
fix data race issues

postindustria-code committed:
extend a/b testing to exitpoint stage, fix empty plan branches, update target value as in spec and other improvements

postindustria-code committed:
use 'no_action' for synthetic HookOutcome, add godocs

postindustria-code committed:
added percentactive validation, deduplicate plan host and account

postindustria-code committed:
add dooh publisher path, fix lint error

postindustria-code committed:
remove per-request warning, add start-up validation

postindustria-code commented:
@antoxas1986  Thank you for the thorough review! I've addressed most of the comments. Below are the ones I left unchanged with my reasoning:

8) Test regression, existing `TestExecuteRawAuctionStage` cases removed
Those cases are still present — the GitHub UI can sometimes make moved or reformatted test cases appear deleted when they aren't. Please check the file directly to confirm.
10) `ABTests` should be an interface
I'd prefer to keep it as a concrete type for now — there's no mocking requirement, no multiple implementations planned, and no cross-package injection, so introducing an interface would add complexity without a clear benefit. Happy to revisit if that changes.
11) `applyABTestPlan` should move to `abtests.go`
Moving `applyABTestPlan` to `abtests.go` would add a `hooks` package dependency to a file that is currently independent. The function is consumed exclusively in `executor.go` and operates on the plan structure, not purely on ABTests state. Its current placement is idiomatic and avoids a layering violation.

Let me know what you think about the updated PR.

postindustria-code committed:
Merge branch 'master' into ab-testing-support-for-modules

antoxas1986 reviewed (changes requested):
## Re-Review: PR #4553 — Modules A/B Testing Support

Most of the prior round's issues have been resolved — the sync.Once fix, stage constants, TOCTOU guard, SetAccountID paths, targeting values, exitpoint wiring, and config validation all look good.

The following issues still need attention before merge:

---

### Medium

**[2] `applyPlan` writes maps without holding `t.mu` — `abtests.go:580–624`**

`applyPlan` is only called from within `initOnce.Do`, and `sync.Once` provides the necessary memory barrier, so this is safe today. But it's fragile: nothing prevents a future refactor from calling `applyPlan` outside `Do`, which would introduce a silent data race. Please either acquire `t.mu.Lock()` at the start of `applyPlan`, or add a clear comment: `// applyPlan must only be called from within initOnce.Do; sync.Once provides the memory barrier for subsequent reads under t.mu.`

**[3] `WriteOutcome` iterates map reference outside lock — `abtests.go:502–555`**

```go
t.mu.RLock()
logMap := t.logMap  // copies map header, not data
runMap := t.runMap
t.mu.RUnlock()
for module, logged := range logMap { ... // outside lock
    if runMap[module] { ...             // outside lock
```
Safe today because maps are only written during `initOnce.Do`. But if `applyPlan` is ever called outside `Do` (e.g. dynamic config reload), this becomes a data race. Recommend iterating under the `RLock`, or copying map contents rather than just the header reference.

**[5] Duplicate analytics tags for modules that ran across multiple stages — `abtests.go:WriteOutcome`**

For *skipped* modules, `checkAndSetLogged` prevents duplicate entries. For modules that *ran* and appear in `outcome.Groups`, there is no such guard — `WriteOutcome` appends a tag every time it is called for any stage that module participated in. If a module runs in multiple stages, it accumulates duplicate tags. If this is intentional per-stage tagging, please document it with a comment.

**[6] Probabilistic path (`PercentActive` 1–99) untested — `abtests.go:623`**

All tests use `PercentActive=0` (always skip) or `PercentActive=100` (always run). The branch `rand.Intn(100) < pa` with 0 < pa < 100 is never exercised. Consider a test with a seeded or injectable rand source, or at least a test that verifies the boundary behaviour.

**[7] Account-level `PercentActive > 100` not validated — `config/hooks.go`**

`Hooks.validate()` checks `HostExecutionPlan` and `DefaultAccountExecutionPlan`. Per-account `config.Account.Hooks.ExecutionPlan.ABTests` is not validated — a value of `PercentActive: 200` will silently pass through and be treated as 100% in `applyPlan`. The `module_code` gap is acknowledged in a comment; the `PercentActive` case is not mentioned.

---

### Minor

**[8] String literal vs constant inconsistency — `executor_test.go`**

Some expected `Status` values use the raw string `"run"` (e.g. lines ~1239, ~1490), others use `hookanalytics.ResultStatusRun`. Use the constant throughout for refactoring safety.

antoxas1986 commented:
You're right on all three points — apologies for the noise.

**Test cases** — Verified directly from the branch (`accfe94`): `TestExecuteProcessedAuctionStage`, `TestExecuteRawAuctionStage`, and `TestExecuteBidderRequestStage` all retain their original cases (mutations, rejection, timeout, etc.). The diff view was misleading due to the large restructuring. Withdrawn.

**ABTests as interface** — Agreed. No cross-package injection, no multiple implementations planned — a concrete type is the right call here.

**applyABTestPlan placement** — Agreed. Moving it to `abtests.go` would pull in the `hooks` package dependency. Its current placement in `executor.go` avoids the layering issue. Accepted.

The remaining open items are still worth addressing before merge:
- [2] `applyPlan` writes maps without `t.mu` — safe due to `sync.Once`, but a comment documenting this invariant would protect future refactors
- [3] `WriteOutcome` iterates a map reference outside the lock — same: safe today, fragile for future changes
- [5] Duplicate analytics tags for modules that ran across multiple stages — please clarify intent with a comment
- [6] Probabilistic `PercentActive` path (1–99) has no test coverage
- [7] Per-account `PercentActive > 100` not validated (only static config is)
- [8] Mix of raw string `"run"` and `hookanalytics.ResultStatusRun` in test assertions

postindustria-code committed:
add explanatory comments and tests

postindustria-code committed:
Merge branch 'master' into ab-testing-support-for-modules

postindustria-code commented:
@antoxas1986 Thank you for the re-review. I have addressed all the comments above, except for №5. There should not be duplicates. `checkAndSetLogged` is called unconditionally for every module where `logMap[module]` is true. I added 2 new UTs to verify that.

antoxas1986 commented:
Hi @postindustria-code — verified at commit `73a0d49`. Quick walk-through of the prior open items:

**[2] `applyPlan` invariant comment** — ✅ Resolved. `hooks/hookexecution/abtests.go:162-165` explicitly documents the `sync.Once` memory barrier.

**[3] `WriteOutcome` map iteration outside RLock** — ✅ Effectively resolved by the same invariant. Minor: the alias-under-RLock pattern at `abtests.go:86-91` reads as if it's snapshotting, but since the maps aren't deep-copied it doesn't actually isolate from a hypothetical writer — the safety comes from the `applyPlan` invariant. A one-line pointer comment at line 89 (e.g. `// safe: maps are immutable post-init; see applyPlan doc`) would help future readers. Non-blocking.

**[5] Duplicate analytics tags across stages** — ✅ Resolved. `TestWriteOutcomeNoDuplicatesForRunModule` (`abtests_test.go:237-309`) exercises the 3-stage case and asserts exactly one tag total; `TestABTestsCheckAndSetLoggedNoDuplicates` covers the 20-goroutine concurrent path. Thanks for adding these.

**[6] Probabilistic `PercentActive` 1–99 coverage** — ✅ Resolved. Injectable `randIntn` (`abtests.go:15`) + `TestPercentActiveBoundaries` (`abtests_test.go:194-235`) covers `pa=1, 50, 99` at both sides of the threshold.

**[8] Raw `"run"`/`"skip"` literals in tests** — Partially resolved. `executor_test.go:5374` (`"hb_ab_module1": "run"`) and `:5395` (`"hb_ab_module1": "skip"`) still use raw strings. Please switch to `string(hookanalytics.ResultStatusRun)` / `ResultStatusSkip` for consistency with the rest of the file.

**New (minor)** — `config/hooks.go:53` doc comment still says: *"The keyword will have value `"1"` if module was run, `"0"` if it was skipped"*. Stale — values are now `"run"`/`"skip"` (`abtests.go:250-253`). Worth updating to avoid confusion.

Apart from the two test-literal lines and the stale comment, this LGTM.
